Enhance checked_objc_cast/dynamic_objc_cast and add is_objc

rniwa opened the pull request:
#### 1b87bf876b4f36380addf41137a13c7aaeb4b04e
<pre>
Enhance checked_objc_cast/dynamic_objc_cast and add is_objc
<a href="https://bugs.webkit.org/show_bug.cgi?id=282811">https://bugs.webkit.org/show_bug.cgi?id=282811</a>

Reviewed by David Kilzer and Timothy Hatcher.

This PR improves checked_objc_cast and dynamic_objc_cast by adding the support for converting U* to T*
and adds is_objc&lt;T&gt;, which like is&lt;T&gt; for Objective-C objects.

Deploy these new functions throughout WebKit codebase.

* Source/WTF/wtf/RetainPtr.h:
(WTF::dynamic_objc_cast): Deleted.
* Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
(WTF::ObjCTypeCastTraits::isType):
(WTF::is_objc):
(WTF::checked_objc_cast):
(WTF::dynamic_objc_cast):
* Source/WTF/wtf/cocoa/UUIDCocoa.mm:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/bridge/objc/objc_utility.mm:
(JSC::Bindings::convertObjcValueToValue):
* Source/WebCore/editing/mac/EditorMac.mm:
* Source/WebCore/platform/cocoa/SerializedPlatformDataCueValue.mm:
(WebCore::SerializedPlatformDataCueValue::SerializedPlatformDataCueValue):
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
(WebCore::isSampleBufferVideoRenderer):
(-[WebAVSampleBufferListenerPrivate layerReadyForDisplayChanged:]):
(-[WebAVSampleBufferListenerPrivate audioRendererWasAutomaticallyFlushed:]):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMFairPlayStreamingAVFObjC.mm:
(WebCore::CDMPrivateFairPlayStreaming::keyIDsForRequest):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::avPlayerLayer const):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::VideoMediaSampleRenderer):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
* Source/WebCore/platform/ios/WidgetIOS.mm:
(WebCore::Widget::convertFromRootToContainingWindow):
(WebCore::Widget::convertFromContainingWindowToRoot):
* Source/WebCore/platform/mac/WidgetMac.mm:
(WebCore::safeRemoveFromSuperview):
(WebCore::Widget::paint):
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::cookieCreated):
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::platformLazyInit):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
* Source/WebCore/rendering/AttachmentLayout.mm:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:_didReceiveInformationalResponse:]):
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::didConnect):
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::actionForMenuItem):
* Source/WebKit/Shared/API/Cocoa/_WKFrameHandle.mm:
(-[_WKFrameHandle isEqual:]):
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::fromNSObject):
* Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
(-[WKNSDictionary objectForKey:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization localizedDictionaryForDictionary:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextInputContext.mm:
(-[_WKTextInputContext isEqual:]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(configureScrollViewWithOverlayRegionsIDs):
* Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
* Source/WebKit/UIProcess/Cocoa/WebPreferencesCocoa.mm:
(WebKit::WebPreferences::platformGetStringUserValueForKey):
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
* Source/WebKit/UIProcess/Gamepad/ios/UIGamepadProviderIOS.mm:
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/Gamepad/mac/UIGamepadProviderMac.mm:
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(existingLocalDragSessionContext):
(ensureLocalDragSessionContext):
* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm:
* Source/WebKit/UIProcess/ios/forms/WKFormColorControl.mm:
(-[WKFormColorControl selectColor:]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.mm:
(-[WKFormSelectControl selectFormPopoverTitle]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(makeResponderFirstResponderIfDescendantOfView):
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebNavigationURLFilter.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
* Tools/DumpRenderTree/mac/UIScriptControllerMac.mm:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TypeCastsCocoa.mm:
(TestWebKitAPI::TEST(TypeCastsCocoa, dynamic_objc_cast)):
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
* Tools/TestWebKitAPI/ios/IOSMouseEventTestHarness.mm:
* Tools/TestWebKitAPI/ios/TestUIMenuBuilder.mm:
* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
* Tools/WebKitTestRunner/ios/UIPasteboardConsistencyEnforcer.mm:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:

Canonical link: <a href="https://commits.webkit.org/286415@main">https://commits.webkit.org/286415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60b4b91e34042aff9b441a4659fc821047c1406c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75934 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80431 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27200 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64105 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59537 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17700 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65211 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39897 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46818 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22695 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25527 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69111 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67933 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23032 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81895 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75210 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3303 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2090 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67764 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3457 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67073 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11021 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97464 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11743 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3253 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6059 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21319 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3274 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4212 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3281 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
     return WTFMove(object);
 }
 
+template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(const RetainPtr<U>& object)
```
You should fix up the `dynamic_objc_cast()` overload in `RetainPtr.h` as well.  Or is there a different plan for that?

(The reason it hasn't moved to `TypeCastsCocoa.h` is that there were uses of it in WebKit and the internal project that required updating `#include` headers to add `#include <wtf/cocoa/TypeCastsCocoa.h>`.)

ddkilzer replied:
The internal project no longer uses `dynamic_objc_cast()` so we only have to fix up WebKit.

Resolving this conversation.

rr-codes commented on Source/WebCore/bridge/objc/objc_utility.mm:
```diff
@@ -195,8 +195,8 @@ JSValue convertObjcValueToValue(JSGlobalObject* lexicalGlobalObject, void* buffe
             id obj = *(const id*)buffer;
-            if ([obj isKindOfClass:[NSString class]])
-                return convertNSStringToString(lexicalGlobalObject, (NSString *)obj);
+            if (auto *str = dynamic_objc_cast<NSString>(obj))
```
shouldn't all of these new `auto *`s be RetainPtr's?

ddkilzer replied:
If we do this, we should retain the object on Line 197/198, then use `.get()` in the `if` statements below:

```
            RetainPtr<id> obj = *(const id*)buffer;
```

This cast just looks wrong—not sure why it's not `(const id)buffer`, and why it doesn't include a `__bridge` attribute with it.  This might also work:

```
            RetainPtr<id> obj = bridge_id_cast(buffer);
```

rniwa replied:
Done that.

ddkilzer commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
+namespace WTF { \
+template <> \
+class ObjCTypeCastTraits<ClassName> { \
+public:
+
+#define SPECIALIZE_OBJC_TYPE_TRAITS_END() \
+}; \
+}
+
+template <typename ExpectedType, typename ArgType>
+inline bool is_objc(ArgType * source)
+{
+    return source && ObjCTypeCastTraits<ExpectedType>::isType(source);
+}
+
+template<typename T> inline T *checked_objc_cast(id object)
+{
+    if (!object)
+        return nullptr;
+
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
+
+    return reinterpret_cast<T*>(object);
+}
+
+template<typename T, typename U> inline T *checked_objc_cast(U *object)
 {
     if (!object)
         return nullptr;
 
-    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION([object isKindOfClass:[T class]]);
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(is_objc<T>(object));
 
     return reinterpret_cast<T*>(object);
 }
 
 // Use dynamic_objc_cast<> instead of checked_objc_cast<> when actively checking NS types,
-// similar to dynamic_cast<> in C++. Be sure to include a nil check.
+// similar to dynamic_cast<> in C++.
```
Why did you decide to remove this text?  I think this is good advice.  (I can't remember if @darinadler advocated for this text or not, back in the day.)

rniwa replied:
Because nil check happens in `dynamic_objc_cast`. There is no point in checking it in the caller side.

ddkilzer commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
     return WTFMove(object);
 }
 
+template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(const RetainPtr<U>& object)
```
[Optional]

Should we have a variant that moves the object to the new `RetainPtr<>` container to avoid ref churn?

Something like this (untested):

```
template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(const RetainPtr<U>&& object)
{
    if (!is_objc<T>(object.get()))
        return nullptr;
    return WTFMove(object);
}
```

If we do this, we should probably write a test that this doesn't cause ref churn under ARC or MRR.  Can do that in `Tools/TestWebKitAPI/Tests/WTF/ns/RetainPtr.mm`.

rniwa replied:
We already have that:
```
template<typename T, typename U> RetainPtr<T> dynamic_objc_cast(RetainPtr<U>&& object)
{
    if (!is_objc<T>(object.get()))
        return nullptr;
    return WTFMove(object);
}
```

ddkilzer commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
 
 using WTF::bridge_cast;
 using WTF::bridge_id_cast;
+using WTF::is_objc;
```
Nit:  This should be last in the list alphabetically.

rniwa replied:
Fixed.

ddkilzer commented on Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
```diff
@@ -421,4 +422,19 @@ SOFT_LINK_CONSTANT_MAY_FAIL_FOR_HEADER(PAL, AVFoundation, AVSampleBufferDisplayL
+SPECIALIZE_OBJC_TYPE_TRAITS_BEGIN(AVSampleBufferAudioRenderer) \
+    static bool isType(id object) { return [object isKindOfClass:PAL::getAVSampleBufferAudioRendererClass()]; }
+    static bool isType(const NSObject *object) { return [object isKindOfClass:PAL::getAVSampleBufferAudioRendererClass()]; }
+SPECIALIZE_OBJC_TYPE_TRAITS_END()
```
You could make this into a single macro that does this, correct?

```
SPECIALIZE_OBJC_TYPE_TRAITS(AVSampleBufferAudioRenderer, PAL::getAVSampleBufferAudioRendererClass())
```

Less boilerplate!  :)

rniwa replied:
Hm... I was trying to match SPECIALIZE_TYPE_TRAITS_BEGIN / SPECIALIZE_TYPE_TRAITS_END but we could do that too.

ddkilzer commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
+
+#define SPECIALIZE_OBJC_TYPE_TRAITS_END() \
+}; \
+}
```
Can we use a single macro here like this (untested)?

```
#define SPECIALIZE_OBJC_TYPE_TRAITS(ClassName, exprReturningClassType) \
namespace WTF { \
template <> \
class ObjCTypeCastTraits<ClassName> { \
public: \
    static bool isType(id object) { return [object isKindOfClass:(exprReturningClassType)]; } \
    static bool isType(const NSObject *object) { return [object isKindOfClass:(exprReturningClassType)]; } \
}; \
}
```

rniwa replied:
I've changed the code to do that.

ddkilzer commented on Source/WebCore/bridge/objc/objc_utility.mm:
```diff
@@ -195,8 +196,8 @@ JSValue convertObjcValueToValue(JSGlobalObject* lexicalGlobalObject, void* buffe
-                return convertNSStringToString(lexicalGlobalObject, (NSString *)obj);
+            if (auto *str = dynamic_objc_cast<NSString>(obj))
+                return convertNSStringToString(lexicalGlobalObject, str);
             if ([obj isKindOfClass:webUndefinedClass()])
```
Can we use `SPECIALIZE_OBJC_TYPE_TRAITS()`here?

```
SPECIALIZE_OBJC_TYPE_TRAITS(WebUndefined, webUndefinedClass())
[...]
            if (auto *webUndef = dynamic_objc_cast<WebUndefined>(objc))
```

May want to put the `SPECIALIZE_OBJC_TYPE_TRAITS(WebUndefined, webUndefinedClass())` definition in `Source/WebCore/bridge/objc/objc_runtime.h`.

rniwa replied:
I don't think it makes much sense here because we won't be using the result of `dynamic_objc_cast`. We could use `is_objc` but I don't think that'll really buy us anything.

ddkilzer commented on Source/WebCore/bridge/objc/objc_utility.mm:
```diff
@@ -195,8 +196,8 @@ JSValue convertObjcValueToValue(JSGlobalObject* lexicalGlobalObject, void* buffe
+                return convertNSStringToString(lexicalGlobalObject, str);
             if ([obj isKindOfClass:webUndefinedClass()])
                 return jsUndefined();
             if ((__bridge CFBooleanRef)obj == kCFBooleanTrue)
```
There are more places that `dynamic_objc_cast<>()` can be deployed below:

```
            if ([obj isKindOfClass:[NSNumber class]])
[...]
            if ([obj isKindOfClass:[NSArray class]])
[...]
            if ([obj isKindOfClass:webScriptObjectClass()]) {
[...]
            if ([obj isKindOfClass:[NSNull class]])
```

rniwa replied:
Those do check `isKindOfClass` but it doesn't cast so there is no need to use `dynamic_objc_cast`.

ddkilzer commented on Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
```diff
@@ -164,7 +165,8 @@ static inline AtomString extractHTTPStatusText(CFHTTPMessageRef messageRef)
 
-        auto messageRef = [m_nsResponse isKindOfClass:[NSHTTPURLResponse class]] ? CFURLResponseGetHTTPResponse([ (NSHTTPURLResponse *)m_nsResponse.get() _CFURLResponse]) : nullptr;
+        auto *urlResponse = dynamic_objc_cast<NSHTTPURLResponse>(m_nsResponse.get());
+        auto messageRef = urlResponse ? CFURLResponseGetHTTPResponse([urlResponse _CFURLResponse]) : nullptr;
```
Do our C++ smart pointer rules allow us to use the `Ref`/`RefPtr` on instance variable to use raw pointers locally?

If not, we should probably use `RetainPtr` here to be consistent.

rniwa replied:
Fixed.

ddkilzer reviewed (approved):
r=me with feedback

Nice cleanup!

rniwa commented on Source/WebCore/bridge/objc/objc_utility.mm:
```diff
@@ -194,19 +195,19 @@ JSValue convertObjcValueToValue(JSGlobalObject* lexicalGlobalObject, void* buffe
-            id obj = *(const id*)buffer;
-            if ([obj isKindOfClass:[NSString class]])
-                return convertNSStringToString(lexicalGlobalObject, (NSString *)obj);
+            RetainPtr<id> obj = bridge_id_cast(buffer);
```
This causes an assertion failure. Reverting this portion:
```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   CoreFoundation                	       0x19727f3f8 CF_IS_OBJC + 56
1   CoreFoundation                	       0x197134b2c CFRetain + 64
2   WebCore                       	       0x137b90ed8 WTF::RetainPtr<objc_object*>::RetainPtr(objc_object*) + 84 (RetainPtr.h:205)
3   WebCore                       	       0x137b90e74 WTF::RetainPtr<objc_object*>::RetainPtr(objc_object*) + 40 (RetainPtr.h:203)
4   WebCore                       	       0x1380bb8d0 JSC::Bindings::convertObjcValueToValue(JSC::JSGlobalObject*, void*, JSC::Bindings::ObjcValueType, JSC::Bindings::RootObject*) + 160 (objc_utility.mm:198)
5   WebCore                       	       0x13f317b58 -[WebScriptObject setValue:forKey:] + 448 (WebScriptObject.mm:410)
6   DumpRenderTree                	       0x104bd8908 -[FrameLoadDelegate didClearWindowObjectInStandardWorldForFrame:] + 236 (FrameLoadDelegate.mm:325)
7   DumpRenderTree                	       0x104bd9100 -[FrameLoadDelegate webView:didClearWindowObjectForFrame:inScriptWorld:] + 96 (FrameLoadDelegate.mm:352)
8   WebKitLegacy                  	       0x10ae46edc objc_object* wtfCallIMP<objc_object*, WebView*, objc_object*, objc_object*>(void (*)(), objc_object*, objc_selector*, WebView*, objc_object*, objc_object*) + 68 (ObjCRuntimeExtras.h:42)
9   WebKitLegacy                  	       0x10ae4505c CallDelegate(void (*)(), WebView*, objc_object*, objc_selector*, objc_object*, objc_object*) + 96 (WebDelegateImplementationCaching.mm:677)
10  WebKitLegacy                  	       0x10ae44ff0 CallFrameLoadDelegate(void (*)(), WebView*, objc_selector*, objc_object*, objc_object*) + 96 (WebDelegateImplementationCaching.mm:1154)
11  WebKitLegacy                  	       0x10acef394 WebFrameLoaderClient::dispatchDidClearWindowObjectInWorld(WebCore::DOMWrapperWorld&) + 184 (WebFrameLoaderClient.mm:1894)
12  WebCore                       	       0x13d3def68 WebCore::FrameLoader::dispatchDidClearWindowObjectInWorld(WebCore::DOMWrapperWorld&) + 248 (FrameLoader.cpp:4541)
13  WebCore                       	       0x13b810a14 WebCore::ScriptController::initScriptForWindowProxy(WebCore::JSWindowProxy&) + 624 (ScriptController.cpp:321)
14  WebCore                       	       0x13b8331e4 WebCore::WindowProxy::createJSWindowProxyWithInitializedScript(WebCore::DOMWrapperWorld&) + 240 (WindowProxy.cpp:143)
15  WebCore                       	       0x13b74601c WebCore::WindowProxy::jsWindowProxy(WebCore::DOMWrapperWorld&) + 120 (WindowProxy.h:73)
16  WebCore                       	       0x13b80f1a4 WebCore::ScriptController::jsWindowProxy(WebCore::DOMWrapperWorld&) + 48 (ScriptController.cpp:432)
17  WebCore                       	       0x13b80ed04 WebCore::ScriptController::evaluateInWorld(WebCore::ScriptSourceCode const&, WebCore::DOMWrapperWorld&) + 204 (ScriptController.cpp:155)
```

rniwa commented on Source/WebKit/Shared/Cocoa/WKObject.h:
```diff
@@ -59,8 +60,7 @@ template<typename WrappedObjectClass> struct WrapperTraits;
 {
-    ASSERT([source isKindOfClass:[DestinationClass class]]);
-    return (DestinationClass *)source;
+    return checked_objc_cast<DestinationClass>(source);
```
Hm... this introduces `isKindOfClass` obj-c message and causes trivial function detector to bail out.
Will revert this portion for now.

rniwa committed:
Enhance checked_objc_cast/dynamic_objc_cast and add is_objc
https://bugs.webkit.org/show_bug.cgi?id=282811

Reviewed by David Kilzer and Timothy Hatcher.

This PR improves checked_objc_cast and dynamic_objc_cast by adding the support for converting U* to T*
and adds is_objc<T>, which like is<T> for Objective-C objects.

Deploy these new functions throughout WebKit codebase.

* Source/WTF/wtf/RetainPtr.h:
(WTF::dynamic_objc_cast): Deleted.
* Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
(WTF::ObjCTypeCastTraits::isType):
(WTF::is_objc):
(WTF::checked_objc_cast):
(WTF::dynamic_objc_cast):
* Source/WTF/wtf/cocoa/UUIDCocoa.mm:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/bridge/objc/objc_utility.mm:
(JSC::Bindings::convertObjcValueToValue):
* Source/WebCore/editing/mac/EditorMac.mm:
* Source/WebCore/platform/cocoa/SerializedPlatformDataCueValue.mm:
(WebCore::SerializedPlatformDataCueValue::SerializedPlatformDataCueValue):
* Source/WebCore/platform/graphics/avfoundation/WebAVSampleBufferListener.mm:
(WebCore::isSampleBufferVideoRenderer):
(-[WebAVSampleBufferListenerPrivate layerReadyForDisplayChanged:]):
(-[WebAVSampleBufferListenerPrivate audioRendererWasAutomaticallyFlushed:]):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMFairPlayStreamingAVFObjC.mm:
(WebCore::CDMPrivateFairPlayStreaming::keyIDsForRequest):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::avPlayerLayer const):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::VideoMediaSampleRenderer):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
* Source/WebCore/platform/ios/WidgetIOS.mm:
(WebCore::Widget::convertFromRootToContainingWindow):
(WebCore::Widget::convertFromContainingWindowToRoot):
* Source/WebCore/platform/mac/WidgetMac.mm:
(WebCore::safeRemoveFromSuperview):
(WebCore::Widget::paint):
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::cookieCreated):
* Source/WebCore/platform/network/cocoa/ResourceResponseCocoa.mm:
(WebCore::ResourceResponse::platformLazyInit):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
* Source/WebCore/rendering/AttachmentLayout.mm:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:_didReceiveInformationalResponse:]):
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
* Source/WebKit/NetworkProcess/cocoa/WebSocketTaskCocoa.mm:
(WebKit::WebSocketTask::didConnect):
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::actionForMenuItem):
* Source/WebKit/Shared/API/Cocoa/_WKFrameHandle.mm:
(-[_WKFrameHandle isEqual:]):
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::fromNSObject):
* Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
(-[WKNSDictionary objectForKey:]):
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization localizedDictionaryForDictionary:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextInputContext.mm:
(-[_WKTextInputContext isEqual:]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(configureScrollViewWithOverlayRegionsIDs):
* Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
* Source/WebKit/UIProcess/Cocoa/WebPreferencesCocoa.mm:
(WebKit::WebPreferences::platformGetStringUserValueForKey):
* Source/WebKit/UIProcess/Cocoa/_WKWarningView.mm:
(colorForItem):
* Source/WebKit/UIProcess/Gamepad/ios/UIGamepadProviderIOS.mm:
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/Gamepad/mac/UIGamepadProviderMac.mm:
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(existingLocalDragSessionContext):
(ensureLocalDragSessionContext):
* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm:
* Source/WebKit/UIProcess/ios/forms/WKFormColorControl.mm:
(-[WKFormColorControl selectColor:]):
* Source/WebKit/UIProcess/ios/forms/WKFormSelectControl.mm:
(-[WKFormSelectControl selectFormPopoverTitle]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(makeResponderFirstResponderIfDescendantOfView):
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebNavigationURLFilter.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
* Tools/DumpRenderTree/mac/UIScriptControllerMac.mm:
* Tools/TestWebKitAPI/Tests/WTF/cocoa/TypeCastsCocoa.mm:
(TestWebKitAPI::TEST(TypeCastsCocoa, dynamic_objc_cast)):
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
* Tools/TestWebKitAPI/Tests/mac/MouseEventTests.mm:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
* Tools/TestWebKitAPI/ios/IOSMouseEventTestHarness.mm:
* Tools/TestWebKitAPI/ios/TestUIMenuBuilder.mm:
* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
* Tools/WebKitTestRunner/ios/UIPasteboardConsistencyEnforcer.mm:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:

Canonical link: https://commits.webkit.org/286415@main

geoffreygaren commented on Source/WTF/wtf/cocoa/TypeCastsCocoa.h:
```diff
@@ -84,31 +84,88 @@ inline RetainPtr<id> bridge_id_cast(RetainPtr<CFTypeRef>&& object)
 
 // Use dynamic_objc_cast<> instead of checked_objc_cast<> when actively checking NS types,
-// similar to dynamic_cast<> in C++. Be sure to include a nil check.
+// similar to dynamic_cast<> in C++.
```
I wish we could match the naming of downcast / dynamicDowncast more closely. Maybe:
is_objc => isObjC
checked_objc_cast => downcastObjC
dynamic_objc_cast => dynamicDowncastObjC

or if we like underscores:
is_objc
checked_objc_cast => downcast_objc
dynamic_objc_cast => dynamic_downcast_objc

That way we have a consistent pattern for describing whether the cast is crashes or returns null.

rniwa replied:
@geoffreygaren : The issue is that these aren't guaranteed to be downcast. It allows "sideway" cast. Maybe:
`isObjC`, `castObjC`, `dynamicCastObjC`?

geoffreygaren replied:
Hmmmm... the goal is still that we should only use these for downcast, right? I guess the challenge is how to static_assert that?

Maybe we can static_assert? This seems to work:

```
@interface Base : NSObject {
}

@end

@interface Derived : Base {
}

@end

void f() {
    static_assert(!std::is_base_of_v<Derived, Base>);
    static_assert(std::is_base_of_v<Base, Derived>);
}

```

rniwa replied:
oh, neat. Didn't realize `is_base_of_v` worked with Obj-C classes.

rniwa replied:
Hm... that doesn't quite work because we have an actual "side-way" cast.

```
/Volumes/Data/safari-2/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/cocoa/TypeCastsCocoa.h:124:19: error: static assertion failed due to requirement 'std::is_base_of_v<id<WKObject>, _WKResourceLoadStatisticsFirstParty>'
  124 |     static_assert(std::is_base_of_v<U, T>);
      |                   ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:27:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdPartyInternal.h:26:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/APIResourceLoadStatisticsThirdParty.h:28:
In file included from /Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/API/APIObject.h:35:
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:63:12: note: in instantiation of function template specialization 'WTF::checked_objc_cast<_WKResourceLoadStatisticsFirstParty, id<WKObject>>' requested here
   63 |     return checked_objc_cast<DestinationClass>(source);
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:68:12: note: in instantiation of function template specialization 'WebKit::checkedObjCCast<_WKResourceLoadStatisticsFirstParty, id<WKObject>>' requested here
   68 |     return checkedObjCCast<typename WrapperTraits<ObjectClass>::WrapperClass>(object.wrapper());
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/Shared/Cocoa/WKObject.h:88:12: note: in instantiation of function template specialization 'WebKit::wrapper<API::ResourceLoadStatisticsFirstParty>' requested here
   88 |     return wrapper(object.get());
      |            ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:51:16: note: in instantiation of function template specialization 'WebKit::wrapper<API::ResourceLoadStatisticsFirstParty>' requested here
   51 |         return wrapper(API::ResourceLoadStatisticsFirstParty::create(data));
      |                ^
/Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:50:12: note: in instantiation of function template specialization 'WTF::createNSArray<const WTF::Vector<WebKit::ITPThirdPartyDataForSpecificFirstParty> &, (lambda at /Volumes/Data/safari-2/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKResourceLoadStatisticsThirdParty.mm:50:60)>' requested here
   50 |     return createNSArray(_thirdParty->underFirstParties(), [] (auto& data) {
      |            ^
1 error generated.
```
